enterprise/server/webhooks/github/github.go

```diff
@@ -137,38 +137,68 @@ func (*githubGitProvider) ParseWebhookData(r *http.Request) (*interfaces.Webhook
 		}, nil
 
 	case *gh.PullRequestEvent:
-		v, err := fieldgetter.ExtractValues(
-			event,
-			"Action",
-			"PullRequest.Base.Ref",
-			"PullRequest.Base.Repo.CloneURL",
-			"PullRequest.Head.Ref",
-			"PullRequest.Head.SHA",
-			"PullRequest.Head.Repo.CloneURL",
-		)
-		if err != nil {
-			return nil, err
-		}
 		// Only build when the PR is opened or pushed to.
-		if !(v["Action"] == "opened" || v["Action"] == "synchronize") {
+		if !(event.GetAction() == "opened" || event.GetAction() == "synchronize") {
 			return nil, nil
 		}
-		isFork := v["PullRequest.Base.Repo.CloneURL"] != v["PullRequest.Head.Repo.CloneURL"]
-		return &interfaces.WebhookData{
-			EventName:     webhook_data.EventName.PullRequest,
-			PushedRepoURL: v["PullRequest.Head.Repo.CloneURL"],
-			PushedBranch:  v["PullRequest.Head.Ref"],
-			SHA:           v["PullRequest.Head.SHA"],
-			TargetRepoURL: v["PullRequest.Base.Repo.CloneURL"],
-			TargetBranch:  v["PullRequest.Base.Ref"],
-			IsTrusted:     !isFork,
-		}, nil
+		return parsePullRequestOrReview(event)
+
+	case *gh.PullRequestReviewEvent:
+		if event.GetAction() != "submitted" {
+			return nil, nil
+		}
+		if event.GetReview().GetState() != "approved" {
+			return nil, nil
+		}
+		if !isTrustedAssociation(event.GetReview().GetAuthorAssociation()) {
+			return nil, nil
+		}
+		wd, err := parsePullRequestOrReview(event)
+		if err != nil {
+			return nil, err
+		}
+		// We now trust the state that the PR is in, since we have an approved review.
+		wd.IsTrusted = true
+		return wd, nil
 
 	default:
 		return nil, nil
 	}
 }
 
+// isTrustedAssociation returns whether a user should be allowed to execute
+// trusted workflows for a repo, given their association to the repo.
+func isTrustedAssociation(association string) bool {
+	return association == "OWNER" || association == "MEMBER" || association == "COLLABORATOR"
+}
+
+// parsePullRequestOrReview extracts WebhookData from a pull_request or
+// pull_request_review event.
+func parsePullRequestOrReview(event interface{}) (*interfaces.WebhookData, error) {
+	v, err := fieldgetter.ExtractValues(
+		event,
+		"Action",
+		"PullRequest.Base.Ref",
+		"PullRequest.Base.Repo.CloneURL",
+		"PullRequest.Head.Ref",
+		"PullRequest.Head.SHA",
+		"PullRequest.Head.Repo.CloneURL",
+	)
+	if err != nil {
+		return nil, err
+	}
+	isFork := v["PullRequest.Base.Repo.CloneURL"] != v["PullRequest.Head.Repo.CloneURL"]
+	return &interfaces.WebhookData{
+		EventName:     webhook_data.EventName.PullRequest,
+		PushedRepoURL: v["PullRequest.Head.Repo.CloneURL"],
+		PushedBranch:  v["PullRequest.Head.Ref"],
+		SHA:           v["PullRequest.Head.SHA"],
+		TargetRepoURL: v["PullRequest.Base.Repo.CloneURL"],
+		TargetBranch:  v["PullRequest.Base.Ref"],
```

Comment:
minor thing, but it would be nice to make the list of fields you're using and referencing match in order. That way a reader can see you're requesting 5 things and then using those same 5, without having to play matchmaker in their head.